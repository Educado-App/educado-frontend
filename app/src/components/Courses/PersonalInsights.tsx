// Interfaces
import { Course } from '../../interfaces/Course';

// Components
import StarRating from "../general/StarRating";

type PersonalInsightsProps = {
  courses: Course[];
}

/**
 * @param props 
 * @param props.courses
 * @returns HTML Element
 * 
 * This component displays personal insights about the user.
 * (Right side of the Courses page at the time of writing)
 */
const PersonalInsights = (props: PersonalInsightsProps) => {
  const courses = props.courses ?? [];

  // TODO: Implement functionality for user info fields
  return (
    <div className='w-full h-full'>
      {/* Welcome message */}
      <h2 className='text-xl font-bold'>Olá User Name</h2>
      {/* Progress section (stats) */}
      <div className='border-y-[1px] border-grayMedium my-4 py-8 '>
        <p className='font-bold'>Progresso</p>
        <div className='grid grid-cols-4 lg:grid-cols-1 gap-y-4'>
          {/* Show stats if the user has > 0 courses */}
          {courses.length ? <>
          <div>
            {/* Total courses */}
            <p>Total cursos</p>
            <p className='font-bold'>8</p>
          </div>
          <div>
            {/* Total students */}
            <p>Total alunos</p>
            <p className='font-bold'>167</p>
          </div>
          <div>
            {/* Total certificates issued */}
            <p>Total certificados emitidos</p>
            <p className='font-bold'>54</p>
          </div>
          <div>
            {/* Rating */}
            <p>Avaliação</p>
            <div className='w-[12rem] max-w-full'>
              <StarRating rating={3.7} className='text-2xl font-bold' />
            </div>
          </div>
          {/* If the user has 0 courses */}
          </> : <p className='italic'>Não há dados suficiente</p> /* There's not enough data */}
        </div>
      </div>
      {/* Activities section */}
      <div className='py-8'>
        <p className='font-bold'>Atividades</p> {/* */}
        <div className='grid grid-cols-4 lg:grid-cols-1 gap-y-2'>
          {/* Show activities if the user has > 0 courses */}
          {courses.length ? <>

          {/* If the user has 0 courses */}
          </> : <p className='italic'>Nenhuma atividade</p> /* No activities */ }
        </div>
      </div>
    </div>
  );
}

export default PersonalInsights;
