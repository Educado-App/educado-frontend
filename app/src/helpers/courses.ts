import { Course } from '../interfaces/Course';

// Gets the total number of subscribers from all courses written by the content creator
export function getTotalSubscriberCount(courses: Course[]) {
	let subscribers = 0;
	courses.forEach(course => {
		subscribers += course.numOfSubscriptions ?? 0;
	});
	return subscribers;
}

// Gets the total number of courses written by the content creator
export function getNumberOfCourses(courses: Course[]) {
	return courses.length;
}

// Gets the average rating of all courses written by the content creator
export function getAverageRating(courses: Course[]) {
	let avgRating = 0;
	courses.forEach(course => {
		avgRating += course.rating ?? 0;
	});
	return avgRating / Math.max(courses.length, 1);
}

// Gets the number of courses with a specific status
export function getCourseCountWithStatus(courses: Course[], status: string) {
	let count = 0;
	courses.forEach(course => {
		if (course.status === status) {
			count++;
		}
	});
	return count;
}
