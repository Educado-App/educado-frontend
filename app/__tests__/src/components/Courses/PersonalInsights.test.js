import renderer from 'react-test-renderer';
import { describe, it, expect } from '@jest/globals'
import { generateTestCourse } from '../../../../__testUtils__/courses/courseTestUtils';

// Components
import PersonalInsights from '../../../../src/components/Courses/PersonalInsights';

// mocks
jest.mock('../../../../src/helpers/userInfo', () => ({
  getUserInfo: jest.fn(() => {
    return {
      id: 1,
      name: 'Test User',
      email: '',
    }
  })
}));


describe('personal insights', () => {
  let courseList = [];
  let component = null;

  beforeAll(async () => {
    // Generate two fake courses
    const course = generateTestCourse();
    course.numOfSubscriptions = 100;
    course.rating = 3;
    const course1 = generateTestCourse();
    course1.numOfSubscriptions = 200;
    course1.rating = 4;

    // Added courses to an array
    courseList[0] = course;
    courseList[1] = course1;

    // Create a test component
    await renderer.act(async () => {
      component = renderer.create(
        <PersonalInsights courses={courseList} />
      );
    });

    
  });

  it('renders correctly', () => {
    // Verify that the component is rendered with snapshot
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('gets total subscriber count from all courses', () => {
    // Verify that the displayed subscriber count is equal to the total of courses
    const totalSubscribers = courseList[0].numOfSubscriptions + courseList[1].numOfSubscriptions;
    expect(component.root.findByProps({ id: 'subscribers' }).children[0]).toBe(totalSubscribers.toString());
  });

  it('correctly displays the amount of courses, when there are no courses', () => {
    // Verify that the displayed subscriber count is equal to the total of courses
    const componentNoCourses = renderer.create(<PersonalInsights courses={[]} />);
    expect(componentNoCourses.root.findByProps({ id: 'noCourses' }).children[0]).toBe("Não há dados suficiente")
  });

  it('gets total amount of courses', () => {
    // Verify that the displayed course amount is equal to the total amount of courses
    expect(component.root.findByProps({ id: 'courseAmount' }).children[0]).toBe(courseList.length.toString());
  });

  it('gets average rating from all courses', () => {
    // Verify that the displayed average rating is equal to the average of courses
    const averageRating = (courseList[0].rating + courseList[1].rating) / courseList.length;
    // expect(component.root.findByProps({ testId: 'averageRating' }).children[0]).toBe(averageRating.toString());
    const starRatingElement = component.root.findByProps({ testId: 'averageRating' });
    // Extract the rating value from the StarRating component
    const displayedRating = parseFloat(starRatingElement.props.rating);
    // Verify that the displayed rating matches the expected average rating
    expect(displayedRating).toBe(averageRating);
  });

});