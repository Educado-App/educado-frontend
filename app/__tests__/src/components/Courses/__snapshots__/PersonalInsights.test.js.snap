// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`personal insights renders correctly 1`] = `
<div
  className="w-full h-full"
>
  <h2
    className="text-xl font-bold"
  >
    Olá 
    Test User
  </h2>
  <div
    className="border-y-[1px] border-grayMedium my-4 py-4"
  >
    <div
      className="grid grid-cols-4 lg:grid-cols-1 gap-y-4"
    >
      <div>
        <div
          className="flex flex-row gap-2"
        >
          <div
            className="w-3 h-3 rounded-full my-auto bg-success"
          />
          <p>
            Publicado
             cursos:
          </p>
          <p
            className="font-bold"
          >
            0
          </p>
        </div>
        <div
          className="flex flex-row gap-2"
        >
          <div
            className="w-3 h-3 rounded-full my-auto bg-[#ffc500]"
          />
          <p>
            Rascunho
             cursos:
          </p>
          <p
            className="font-bold"
          >
            2
          </p>
        </div>
        <div
          className="flex flex-row gap-2"
        >
          <div
            className="w-3 h-3 rounded-full my-auto bg-grayMedium"
          />
          <p>
            Oculto
             cursos:
          </p>
          <p
            className="font-bold"
          >
            0
          </p>
        </div>
        <div
          className="flex flex-row gap-2 ml-5"
        >
          <p>
            Total cursos:
          </p>
          <p
            className="font-bold"
            id="courseAmount"
          >
            2
          </p>
        </div>
      </div>
      <div>
        <p>
          Total alunos
        </p>
        <p
          className="font-bold"
          id="subscribers"
        >
          300
        </p>
      </div>
      <div>
        <p>
          Avaliação
        </p>
        <div
          className="w-[12rem] max-w-full"
        >
          <div
            className="flex flex-row"
          >
            <svg
              className="text-star"
              role="presentation"
              style={{}}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                style={
                  {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
            <svg
              className="text-star"
              role="presentation"
              style={{}}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                style={
                  {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
            <svg
              className="text-star"
              role="presentation"
              style={{}}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,17.27L18.18,21L16.54,13.97L22,9.24L14.81,8.62L12,2L9.19,8.62L2,9.24L7.45,13.97L5.82,21L12,17.27Z"
                style={
                  {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
            <svg
              className="text-star"
              role="presentation"
              style={{}}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,15.4V6.1L13.71,10.13L18.09,10.5L14.77,13.39L15.76,17.67M22,9.24L14.81,8.63L12,2L9.19,8.63L2,9.24L7.45,13.97L5.82,21L12,17.27L18.18,21L16.54,13.97L22,9.24Z"
                style={
                  {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
            <svg
              className="text-star"
              role="presentation"
              style={{}}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,15.39L8.24,17.66L9.23,13.38L5.91,10.5L10.29,10.13L12,6.09L13.71,10.13L18.09,10.5L14.77,13.38L15.76,17.66M22,9.24L14.81,8.63L12,2L9.19,8.63L2,9.24L7.45,13.97L5.82,21L12,17.27L18.18,21L16.54,13.97L22,9.24Z"
                style={
                  {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
            <p
              className="text-star ml-1 text-2xl font-bold"
            >
              3.5
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <p
      className="font-bold"
    >
      Atividades
    </p>
     
    <div
      className="grid grid-cols-4 lg:grid-cols-1 gap-y-2"
    />
  </div>
</div>
`;
